### PR TITLE
cargo-install: prefer building artifacts in the system temporary directory

### LINKED ISSUE #2606: cargo-install: build directory should be in $TEMPDIR
and each instance of `cargo-install` should have its own build directory.

Right now `cargo-install` uses a `target-install` directory in `$PWD`. So you run into issues like these:

You don't have permission to write to `$PWD`:

```
$ cd /
$ cargo install cargo-graph
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading cargo-graph v0.2.2
error: failed to compile `cargo-graph v0.2.2`, intermediate artifacts can be found at `/target-install`

Caused by:
  failed to open: /target-install/release/.cargo-lock

To learn more, run the command again with --verbose.
```

Calling `cargo install` multiple times from the same directory will build each crate sequentially instead of in parallel (because the build directory gets locked -- without the lock things would be worse, so removing the lock is not a solution)

```
$ cargo install cargo-graph & cargo install cargo-count
[1] 28438
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Blocking waiting for file lock on the registry index
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Blocking waiting for file lock on the registry index
    Blocking waiting for file lock on crate metadata
    Blocking waiting for file lock on build directory
   Compiling rustc-serialize v0.3.19
(...)
```

If the build directory was `$(mktemp -d)` then you wouldn't have problem calling `cargo install` from a directory you don't permission to write to, and you could spawn several `cargo install` (from the same directory) that actually run in parallel.

Thoughts?

### PR DESCRIPTION
**Pull request by japaric**
and each cargo-install instance creates and uses its own build directory. This
allows running several cargo-install instances in parallel.

If we fail to create a temporary directory for whatever reason fallback to
creating and using a target-install directory in the current directory.

closes #2606

---

r? @alexcrichton 

Qs:
- Should we preserve the current behavior (`target-install` in `cwd`) as a fallback or remove it and error if we can't create a `TempDir` in `env::temp_dir()`? (we currently error if we can't create `target-install` directory in `cwd`)

- Should I add tests for the issues I raised at #2606? If yes, how can I test `cargo-install` parallelism? Lack of "Blocking waiting for file lock on build directory" in the output of the `cargo` commands? or something else?

### CODE REVIEW

File: src/cargo/ops/cargo_install.rs [hunk 1]
```diff
@@ -79,13 +80,25 @@ pub fn install(root: Option<&str>,
         try!(check_overwrites(&dst, &pkg, &opts.filter, &list));
     }
 
+    let mut td_opt = None;
     let target_dir = if source_id.is_path() {
         config.target_dir(&pkg)
     } else {
-        Filesystem::new(config.cwd().join("target-install"))
+        if let Ok(td) = TempDir::new("cargo-install") {
```
Comment (japaric) [inline src/cargo/ops/cargo_install.rs]: Should this be `target-install` instead of `cargo-install` for consistency with the current/fallback behavior?

Comment (alexcrichton) [inline src/cargo/ops/cargo_install.rs]: Nah this is fine

**Comment by alexcrichton** [comment]
Yeah probably can't hurt to add tests. You can check out some tests in `tests/test_cargo_concurrent.rs` and perhaps do something similar?

**Comment by japaric** [comment]
Pushed a commit adding tests. Checked that both tests don't pass without this PR.

**Pull request merged**